format-xls/src/main/java/io/cdap/plugin/format/xls/input/XlsRowConverter.java

```diff
@@ -0,0 +1,167 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.xls.input;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.DataFormatter;
+import org.apache.poi.ss.usermodel.DateUtil;
+import org.apache.poi.ss.usermodel.FormulaEvaluator;
+import org.apache.poi.ss.usermodel.Row;
+
+import java.util.List;
+
+/**
+ * Converts a row of XLS cells to a StructuredRecord.
+ */
+public class XlsRowConverter {
+  private final FormulaEvaluator evaluator;
+  private static final DataFormatter dataFormatter = new DataFormatter();
+
+  XlsRowConverter(FormulaEvaluator evaluator) {
+    this.evaluator = evaluator;
+  }
+
+  /**
+   * Converts a row of XLS cells to a StructuredRecord.
+   * Returns null if the row is null or empty.
+   */
+  public StructuredRecord.Builder convert(Row row, Schema outputSchema) {
+    if (row == null) {
+      return null;
+    }
+    boolean isRowEmpty = true;
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    List<Schema.Field> fields = outputSchema.getFields();
+    for (int cellIndex = 0; cellIndex < row.getLastCellNum(); cellIndex++) {
+      if (cellIndex >= fields.size()) {
+        throw new IllegalArgumentException(
+                String.format("Schema contains fewer fields than the number of columns in the excel file. " +
+                        "Schema fields: %s, Excel columns: %s", fields.size(), row.getLastCellNum()));
+      }
+      Cell cell = row.getCell(cellIndex, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+      if (cell == null) {
+        // Blank cells are skipped, builder will set null for the field, no processing needed.
+        continue;
+      }
+      Schema.Field field = fields.get(cellIndex);
+      Schema.Type type = field.getSchema().isNullable() ?
+              field.getSchema().getNonNullable().getType() : field.getSchema().getType();
+      Object cellValue;
+      switch (type) {
+        case STRING:
+          cellValue = getCellAsString(cell);
+          break;
+        case DOUBLE:
+          cellValue = getCellAsDouble(cell);
+          break;
+        case BOOLEAN:
+          cellValue = getCellAsBoolean(cell);
+          break;
+        default:
+          throw new IllegalArgumentException(
+                  String.format("Field '%s' is of unsupported type '%s'. Supported types are: %s",
+                          field.getName(), type, "string, double, boolean"));
+      }
+      if (cellValue == null) {
+        continue;
+      }
+      builder.set(field.getName(), cellValue);
+      isRowEmpty = false;
+    }
+    if (isRowEmpty) {
+      return null;
```

Comment:
method should be annotated with `@Nullable` so it is clear the caller needs to handle nulls.

Reply:
Marked !